controllers/deactivation/deactivation_controller.go

```diff
@@ -199,6 +200,14 @@ func (r *Reconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.
 		// It is not yet time to deactivate so requeue when it will be
 		requeueAfterExpired := time.Until(deactivationDueTime)
 
+		if usersignup.Status.ScheduledDeactivationTimestamp.IsZero() || !usersignup.Status.ScheduledDeactivationTimestamp.Time.Equal(deactivationDueTime) {
+			usersignup.Status.ScheduledDeactivationTimestamp = v1.NewTime(deactivationDueTime)
+			if err := r.Client.Status().Update(ctx, usersignup); err != nil {
```

Comment:
Update of the UserSignup status should be managed by UserSignup controller, not by deactivation controller. Deactivation controller should manage only the spec of the UserSignup resource.

Reply:
Ideally yes, but how do you suggest to make the UserSignup controller to set the `scheduledDeactivationTimestamp` status? I see two options:
- Rewrite our deactivation mechanism and introduce a new spec in UserSignup, which would tell the UserSignup controller when to deactivate instead of using our current "state" filed.
- Move the deactivation timestamp calculation from the deactivation controller to the UserSignup controller which would mean code duplication because the deactivation controller already does this job.

IMO adding the UserSignup status thing to the deactivation controller is a better option. Or you are thinking about something else?

Reply:
We could have used an annotation instead. That would be totally OK, right?

Reply:
as for the second option - which code duplication are you talking about? You can easily extract the logic that calculates the deactivation time based on the MUR.spec and call it from both places. In the case of usersignup, you could call it here: https://github.com/codeready-toolchain/host-operator/blob/188cb8ad8108505af07486656e3bbbdd8dd27615/controllers/usersignup/usersignup_controller.go#L382-L383
at this point you know that the user is approved, ready to be used, you have the latest MUR, and you know that when MUR is updated then you get to this line to reset the value in the status.

Reply:
apart from that, the logic that is currently in the PR won't work - it would set the value only when the UserSignup is in the deactivating state, not when it's before the "deactivating" time (it calculates the `deactivationDueTime` based on the LastTransitionTime of the `UserSignupUserDeactivatingNotificationCreated` condition.
The controller exists earlier for the case when it's not in the "deactivating" state.

Reply:
> apart from that, the logic that is currently in the PR won't work

This is why we need tests to prove that it works in all expected cases :)

I don't really like the annotation approach but we could extract the logic of calculating the deactivation timestamp indeed.

Reply:
Logic for setting the scheduled deactivation time is now in the UserSignup controller.  I'm still uncertain whether we should be depending on this timestamp value in the deactivation controller itself, it currently recalculates everything independently and covers additional stuff like the deactivation notification calculation, so I'm tempted to just leave it as is.

Reply:
The information about account deactivation is pretty critical. I don't like the idea of having two completely independent logics, one for the user's information and one for the actual deactivation. I like the idea of moving the calculation to the UserSignup controller while the deactivation controller uses this tiemstamp from the status instead of re-calculating it on it's own.

Reply:
> Logic for setting the scheduled deactivation time is now in the UserSignup controller

I don't see it in this PR. Make sure you didn't forget to update the PR ;)

Reply:
@alexeykazakov unfortunately it seems we have no choice about having two separate sources of logic for the deactivation.  If you recall, we were quite strict in our design with ensuring that the deactivation "countdown" only began once we notified the user that their account would be deactivated shortly, generally 3 days I believe before deactivation.  We contended with the possibility that for whatever reason, the user might not be notified due to unforeseen technical issues and due to this possibility their trial may be extended slightly as a result due to delayed issuing of their deactivation notification.  

If I recall correctly, this was to allow the user sufficient opportunity to make whatever preparations they deem necessary (backup or save whatever state they may have in their sandbox) before deactivation occurs.  This requirement unfortunately creates a situation where the actual deactivation time is at odds with the scheduled deactivation time which is calculated at the start of the trial period.  This is why I believe our best option might be to leave the scheduled deactivation time in the UserSignup as informational state only, and to continue having the deactivation controller calculate the actual deactivation time based on when the notification is created.

Reply:
But in this case we will have incorrect deactivation timestamp in the UserSignup. If the deactivating notification was not created at least 3 days ago even if the deactivation is well past due then the deactivation timestamp will point to the past while the actual deactivation is still pending.
I looked at the deactivation controller more closely and I see there is many things which dictate when the user signup is actually going to be deactivated. Tier settings, deactivating notification status, if the user's email is in the list of the exceptions, etc. So I don't see how we can easily duplicate this logic in the UserSignup controller without effectively duplicating almost entire deactivation controller there. I also don't see any efficient way of extracting this logic to some common code which we could re-use in both controller.
So I think keeping this logic in the deactivation controller (which purpose is entirely to know when deactivation is supposed to happen) exclusively is the best approach. And having to update the UserSignup status (the new deactivation timestamp field) from the deactivation controller is relatively small price to pay.